gridcomps/History/MAPL_StationSamplerMod.F90

```diff
@@ -0,0 +1,461 @@
+  use LocStreamFactoryMod
+  use pFIO
+  use MAPL_TimeDataMod
+  use MAPL_VerticalDataMod
+  use MAPL_BaseMod
+  use MAPL_CommsMod
+  use MAPL_LocstreamRegridderMod
+  use, intrinsic :: iso_fortran_env, only: REAL32
+  use, intrinsic :: iso_fortran_env, only: REAL64
+  implicit none
+  private
+
+  public :: StationSampler
+  type :: StationSampler
+     type(LocStreamFactory) :: LSF
+     type(ESMF_LocStream) :: esmf_ls 
+     type(LocstreamRegridder) :: regridder
+     type(ESMF_time), allocatable :: times(:)
+     integer, allocatable :: ids(:)
+     integer :: nobs
+     integer :: nstation
+     integer, allocatable :: station_id(:)
+     character(len=ESMF_MAXSTR), allocatable :: station_name(:)
+     real(kind=REAL64), allocatable :: lons(:)
+     real(kind=REAL64), allocatable :: lats(:)
+     real(kind=REAL64), allocatable :: elevs(:)
+     !
+     type(ESMF_FieldBundle) :: bundle
+     type(FileMetadata) :: fmd
+     type(NetCDF4_FileFormatter) :: formatter
+     type(VerticalData) :: vdata
+     logical :: do_vertical_regrid
+     type(TimeData) :: time_info
+     integer :: previous_index
+     integer :: obs_written  ! replace number_written
+     character(LEN=ESMF_MAXPATHLEN) :: ofile ! file_name
+   contains
+     procedure :: add_metadata_route_handle
+     procedure :: create_file_handle
+     procedure :: close_file_handle
+     procedure :: append_file
+     procedure :: get_file_start_time
+     procedure :: compute_time_for_current
+     !!procedure :: destroy_sampler  !! destructor __ deallocate arrays
+  end type StationSampler
+
+  interface StationSampler
+     module procedure new_StationSampler_readfile
+  end interface StationSampler
+  integer :: maxstr = 2048  ! because ESMF_MAXSTR=256  
+contains
+
+
+  function new_StationSampler_readfile (filename,rc) result(sampler)  
+    type(StationSampler) :: sampler
+    character(len=*), intent(in) :: filename ! 1:station_name input
+    integer, optional, intent(out) :: rc
+    ! loc
+    character(len=40) :: str, sdmy, shms
+    integer :: max_len, max_seg, nseg
+    integer :: unit, ios, nline, id, nstation, status, i, j, nobs
+    integer :: iday, imonth, iyear, ihr, imin, isec
+    real :: x, y, z, t
+    character(len=1) :: s1
+    
+    !__ 1. read from station_id_file: static
+    !      plain text format: [id,name,lat,lon,elev]
+    !
+    write(6,*) 'sampler: station_id filename= ', trim(filename)
+    open(newunit=unit, file=trim(filename), form='formatted', access='sequential', status='old')
+    ios=0; nstation=0
+    do while (ios==0)
+       read (unit, *, IOSTAT=ios)  id, str, x, y, z
+       if (ios==0) nstation=nstation+1
+       !! write(6,*) 'id=',id
+    enddo
+    !!write(6,*) 'nstation=', nstation
+    sampler%nstation=nstation
+    allocate(sampler%station_id(nstation))
+    allocate(sampler%station_name(nstation))
+    allocate(sampler%lons(nstation))
+    allocate(sampler%lats(nstation))
+    allocate(sampler%elevs(nstation))
```

Comment:
Will become single ALLOCATE statement with the derived type.

Reply:
These array are used in subsequent subroutine calls.